src/TraceEvent/TraceUtilities/HistoryDictionary.cs

```diff
@@ -25,52 +25,89 @@ public HistoryDictionary(int initialSize)
+
+                // See if we can jump ahead.  Currently we only do this of the first entry, 
+                // But you could imagine using some of the other nodes's skipAhead entries.   
+                if (firstEntry.skipAhead != null && firstEntry.skipAhead.startTime <= startTime)
+                    entry = firstEntry.skipAhead;
+
+                for (; ; )
                 {
-                    entry[startTime] = value;
+                    // We found exact match
+                    if (startTime == entry.StartTime)
+                    {
+                        // Just update the value and exit immediately as there is no need to
+                        // update skipAhead or increment count
+                        entry.value = value;
+                        return;
+                    }
+
+                    if (entry.next == null)
+                    {
+                        entry.next = new HistoryValue(startTime, id, value);
+                        break;
+                    }
+
+                    // We sort the entries from smallest to largest time. 
+                    if (startTime < entry.startTime)
+                    {
+                        // This entry belongs in front of this entry.  
+                        // Insert it before the current entry by moving the current entry after it.  
+                        HistoryValue newEntry = new HistoryValue(entry);
+                        entry.startTime = startTime;
+                        entry.value = value;
+                        entry.next = newEntry;
+                        Debug.Assert(entry.startTime <= entry.next.startTime);
+                        break;
+                    }
+                    entry = entry.next;
                 }
+                firstEntry.skipAhead = entry.next;
             }
+            count++;
         }
         // TryGetValue will return the value associated with an id that was placed in the stream 
         // at time100ns OR BEFORE.  
         public bool TryGetValue(Address id, long time, out T value)
         {
-            SortedDictionary<long, T> entry;
+            HistoryValue entry;
             if (entries.TryGetValue((long)id, out entry))
             {
                 // The entries are sorted smallest to largest.  
                 // We want the last entry that is smaller (or equal) to the target time) 
-                long? lastStart = null;
-                var keys = entry.Keys;
-                foreach (var t in keys)
+
+                var firstEntry = entry;
+                // See if we can jump ahead.  Currently we only do this of the first entry, 
+                // But you could imagine using some of the other nodes's skipAhead entries.   
+                if (firstEntry.skipAhead != null && firstEntry.skipAhead.startTime < time)
+                    entry = firstEntry.skipAhead;
+
+                HistoryValue last = null;
+                for (; ; )
```

Comment:
Latest version does not have this change (I just kept the original code). Should I change it to `while (true)`?